syntax = "proto3";
import "common.proto";

package rpc_general;

service AsyncInterface {
    rpc AsyncInterface_Init (AIInitParams) returns (Status) {}

    rpc AsyncInterface_Enable (AIInterfaceId) returns (Status) {}
    rpc AsyncInterface_Disable (AIInterfaceId) returns (Status) {}

    rpc AsyncInterface_ReadReq (AIReadReqParams) returns (Status) {}
    rpc AsyncInterface_GetReadData (AIInterfaceId) returns (AIGetReadDataReturn) {}

    rpc AsyncInterface_WriteReq (AIWriteReqParams) returns (Status) {}
}

enum AIIrqFlags {
    AIIrqFlags_RX_FULL = 0;
    AIIrqFlags_TX_EMPTY = 1;
    AIIrqFlags_END_TRANSFER = 2;
    AIIrqFlags_ERROR = 3;
    AIIrqFlags_RX_IDLE_LINE = 4;
}

message AIInitParams {
    uint32 interface_id = 1;
    int32 irq_id = 2;
}

message AIInterfaceId {
    uint32 interface_id = 1;
}

message AIReadReqParams {
    uint32 interface_id = 1;
    uint32 data_size = 2;
}

message AIGetReadDataReturn {
    Status status = 1;
    bytes data_bytes = 2;
}

message AIWriteReqParams {
    uint32 interface_id = 1;
    bytes data_bytes = 2;
}
